pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3407 @@
+        yield v
+
+
+# =====================================================
+#  identify_variables
+# =====================================================
+
+class _VariableVisitor(SimpleExpressionVisitor):
+
+    def __init__(self):
+        self.seen = set()
+
+    def visit(self, node):
+        if node.__class__ in nonpyomo_leaf_types:
+            return
+
+        if node.is_variable_type():
+            if id(node) in self.seen:
+                return
+            self.seen.add(id(node))
+            return node
+
+
+def identify_variables(expr, include_fixed=True):
+    """
+    A generator that yields a sequence of variables
+    in an expression tree.
+
+    Args:
+        expr: The root node of an expression tree.
+        include_fixed (bool): If :const:`True`, then
+            this generator will yield variables whose
+            value is fixed.  Defaults to :const:`True`.
+
+    Yields:
+        Each variable that is found.
+    """
+    visitor = _VariableVisitor()
+    if include_fixed:
+        for v in visitor.xbfs_yield_leaves(expr):
+            yield v
+    else:
+        for v in visitor.xbfs_yield_leaves(expr):
+            if not v.is_fixed():
+                yield v
+
+
+# =====================================================
+#  identify_mutable_parameters
+# =====================================================
+
+class _MutableParamVisitor(SimpleExpressionVisitor):
+
+    def __init__(self):
+        self.seen = set()
+
+    def visit(self, node):
+        if node.__class__ in nonpyomo_leaf_types:
+            return
+
+        # TODO: Confirm that this has the right semantics
+        if not node.is_variable_type() and node.is_fixed():
+            if id(node) in self.seen:
+                return
+            self.seen.add(id(node))
+            return node
+
+
+def identify_mutable_parameters(expr):
+    """
+    A generator that yields a sequence of mutable
+    parameters in an expression tree.
+
+    Args:
+        expr: The root node of an expression tree.
+
+    Yields:
+        Each mutable parameter that is found.
+    """
+    visitor = _MutableParamVisitor()
+    for v in visitor.xbfs_yield_leaves(expr):
+        yield v
+
+
+# =====================================================
+#  _polynomial_degree
+# =====================================================
+
+class _PolynomialDegreeVisitor(ExpressionValueVisitor):
+
+    def visit(self, node, values):
+        """ Visit nodes that have been expanded """
+        return node._compute_polynomial_degree(values)
+
+    def visiting_potential_leaf(self, node):
+        """
+        Visiting a potential leaf.
+
+        Return True if the node is not expanded.
+        """
+        if node.__class__ in nonpyomo_leaf_types or not node.is_potentially_variable():
+            return True, 0
+
+        if not node.is_expression_type():
+            return True, 0 if node.is_fixed() else 1
+
+        return False, None
+
+
+def _polynomial_degree(node):
```

Comment:
This "helper" function isn't used anywhere in Pyomo except one place in this file.  We should remove it.